File: crates/openwilly-player/src/assets/bitmap.rs
//! Bitmap decoder for Director BITD chunks
//!
//! Supports:
//! - 1-bit bitmaps (bitfield)
//! - 8-bit uncompressed (palette-indexed, 0xFF-inverted)
//! - 8-bit PackBits RLE (palette-indexed, 0xFF-inverted)
//! - 16-bit uncompressed/RLE
//! - 32-bit planar RLE (A,R,G,B channels, no inversion)

/// Decoded bitmap in RGBA format, ready for rendering
#[derive(Debug, Clone)]
pub struct DecodedBitmap {
    pub width: u32,
    pub height: u32,
    pub pixels: Vec<u8>, // RGBA, 4 bytes per pixel
}

/// Decode a Director BITD chunk into an RGBA bitmap
///
/// `transparent_color`: which palette index should be fully transparent.
///   - `Some(255)` for overlay sprites (Director default: white = transparent)
///   - `None` for backgrounds (no transparency)
pub fn decode_bitd(
    data: &[u8],
    width: u16,
    height: u16,
    bit_depth: u8,
    palette: &[[u8; 3]],
    transparent_color: Option<u8>,
) -> DecodedBitmap {
    let w = width as u32;
    let h = height as u32;
    let mut pixels = vec![0u8; (w * h * 4) as usize];

    match bit_depth {
        1 => decode_1bit(data, w, h, &mut pixels),
        8 => decode_8bit(data, w, h, palette, transparent_color, &mut pixels),
        16 => decode_16bit(data, w, h, &mut pixels),
        32 => decode_32bit(data, w, h, &mut pixels),
        _ => {
            tracing::warn!("Unsupported bit depth: {}", bit_depth);
        }
    }

    DecodedBitmap {
        width: w,
        height: h,
        pixels,
    }
}

// ============================================================================
// 1-bit decode
// ============================================================================

fn decode_1bit(data: &[u8], w: u32, h: u32, pixels: &mut [u8]) {
    // Row stride: ceil(width / 16) * 2 bytes (16-bit aligned)
    let row_bytes = ((w + 15) / 16 * 2) as usize;

    for y in 0..h as usize {
        for x in 0..w as usize {
            let byte_idx = y * row_bytes + x / 8;
            let bit_idx = 7 - (x % 8);
            let bit = if byte_idx < data.len() {
                (data[byte_idx] >> bit_idx) & 1
            } else {
                0
            };
            // 1-bit: 0 = white, 1 = black (color inverted)
            let color = if bit == 0 { 255u8 } else { 0u8 };
            let px = (y as u32 * w + x as u32) as usize * 4;
            if px + 3 < pixels.len() {
                pixels[px] = color;
                pixels[px + 1] = color;
                pixels[px + 2] = color;
                pixels[px + 3] = 255;
            }
        }
    }
}

// ============================================================================
// 8-bit decode (uncompressed + PackBits RLE)
// ============================================================================

fn decode_8bit(data: &[u8], w: u32, h: u32, palette: &[[u8; 3]], transparent_color: Option<u8>, pixels: &mut [u8]) {
    // Row stride: width padded to even
    let row_bytes = if w % 2 == 0 { w } else { w + 1 } as usize;
    let expected_uncompressed = row_bytes * h as usize;

    if data.len() >= expected_uncompressed {
        // Uncompressed
        decode_8bit_uncompressed(data, w, h, row_bytes, palette, transparent_color, pixels);
    } else {
        // PackBits RLE
        decode_8bit_rle(data, w, h, row_bytes, palette, transparent_color, pixels);
    }
}

fn decode_8bit_uncompressed(
    data: &[u8],
    w: u32,
    h: u32,
    row_bytes: usize,
    palette: &[[u8; 3]],
    transparent_color: Option<u8>,
    pixels: &mut [u8],
) {
    for y in 0..h as usize {
        for x in 0..w as usize {
            let idx = y * row_bytes + x;
            if idx < data.len() {
                let color_idx = 0xFF ^ data[idx]; // Director inverts with 0xFF XOR
                let px_offset = (y as u32 * w + x as u32) as usize * 4;
                if px_offset + 3 < pixels.len() && (color_idx as usize) < palette.len() {
                    let c = palette[color_idx as usize];
                    pixels[px_offset] = c[0];
                    pixels[px_offset + 1] = c[1];
                    pixels[px_offset + 2] = c[2];
                    // Transparent if matches the designated background color
                    pixels[px_offset + 3] = match transparent_color {
                        Some(tc) if color_idx == tc => 0,
                        _ => 255,
                    };
                }
            }
        }
    }
}

fn decode_8bit_rle(
    data: &[u8],
    w: u32,
    h: u32,
    row_bytes: usize,
    palette: &[[u8; 3]],
    transparent_color: Option<u8>,
    pixels: &mut [u8],
) {
    // PackBits RLE decompression
    // Decompress entire data into row-aligned buffer, then apply palette
    let mut decompressed = vec![0u8; row_bytes * h as usize];
    let mut src = 0usize;
    let mut dst = 0usize;

    while src < data.len() && dst < decompressed.len() {
        let run_len = data[src] as i8;
        src += 1;

        if run_len >= 0 {
            // Literal run: copy next (run_len + 1) bytes
            let count = run_len as usize + 1;
            for _ in 0..count {
                if src < data.len() && dst < decompressed.len() {
                    decompressed[dst] = data[src];
                    src += 1;
                    dst += 1;
                }
            }
        } else if run_len as u8 != 0x80 {
            // Repeat run: repeat next byte (1 - run_len) times
            // run_len is negative, so (1 - run_len) = (1 + |run_len|)
            let count = (1i16 - run_len as i16) as usize;
            let val = if src < data.len() {
                let v = data[src];
                src += 1;
                v
            } else {
                0
            };
            for _ in 0..count {
                if dst < decompressed.len() {
                    decompressed[dst] = val;
                    dst += 1;
                }
            }
        }
        // 0x80 = no-op
    }

    // Apply palette from decompressed buffer
    decode_8bit_uncompressed(&decompressed, w, h, row_bytes, palette, transparent_color, pixels);
}

// ============================================================================
// 16-bit decode
// ============================================================================

fn decode_16bit(data: &[u8], w: u32, h: u32, pixels: &mut [u8]) {
    let row_bytes = w as usize * 2;
    let expected_uncompressed = row_bytes * h as usize;

    let decompressed = if data.len() >= expected_uncompressed {
        data.to_vec()
    } else {
        packbits_decompress(data, row_bytes * h as usize)
    };

    for y in 0..h as usize {
        for x in 0..w as usize {
            let idx = (y * w as usize + x) * 2;
            if idx + 1 < decompressed.len() {
                let pixel16 = (decompressed[idx] as u16) << 8 | decompressed[idx + 1] as u16;
                let r = ((pixel16 >> 10) & 0x1F) as u8;
                let g = ((pixel16 >> 5) & 0x1F) as u8;
                let b = (pixel16 & 0x1F) as u8;
                let px_offset = (y as u32 * w + x as u32) as usize * 4;
                if px_offset + 3 < pixels.len() {
                    pixels[px_offset] = (r << 3) | (r >> 2);
                    pixels[px_offset + 1] = (g << 3) | (g >> 2);
                    pixels[px_offset + 2] = (b << 3) | (b >> 2);
                    pixels[px_offset + 3] = 255;
                }
            }
        }
    }
}

// ============================================================================
// 32-bit decode (planar ARGB)
// ============================================================================

fn decode_32bit(data: &[u8], w: u32, h: u32, pixels: &mut [u8]) {
    // 32-bit bitmaps use planar channel layout:
    // Each row is stored as: [A0, A1, ..., An, R0, R1, ..., Rn, G0, ..., Gn, B0, ..., Bn]
    // No color inversion for 32-bit

    let row_bytes = w as usize * 4; // 4 channels per row
    let total = row_bytes * h as usize;

    let decompressed = if data.len() >= total {
        data.to_vec()
    } else {
        packbits_decompress(data, total)
    };

    for y in 0..h as usize {
        let row_start = y * row_bytes;
        let channel_stride = w as usize;
        for x in 0..w as usize {
            let a_idx = row_start + x;
            let r_idx = row_start + channel_stride + x;
            let g_idx = row_start + channel_stride * 2 + x;
            let b_idx = row_start + channel_stride * 3 + x;

            let px_offset = (y as u32 * w + x as u32) as usize * 4;
            if px_offset + 3 < pixels.len() && b_idx < decompressed.len() {
                pixels[px_offset] = decompressed[r_idx];
                pixels[px_offset + 1] = decompressed[g_idx];
                pixels[px_offset + 2] = decompressed[b_idx];
                pixels[px_offset + 3] = decompressed[a_idx];
            }
        }
    }
}

// ============================================================================
// PackBits decompression (generic)
// ============================================================================

fn packbits_decompress(data: &[u8], expected_len: usize) -> Vec<u8> {
    let mut out = Vec::with_capacity(expected_len);
    let mut src = 0usize;

    while src < data.len() && out.len() < expected_len {
        let run_len = data[src] as i8;
        src += 1;

        if run_len >= 0 {
            let count = run_len as usize + 1;
            for _ in 0..count {
                if src < data.len() && out.len() < expected_len {
                    out.push(data[src]);
                    src += 1;
                }
            }
        } else if run_len as u8 != 0x80 {
            let count = (1i16 - run_len as i16) as usize;
            let val = if src < data.len() {
                let v = data[src];
                src += 1;
                v
            } else {
                0
            };
            for _ in 0..count {
                if out.len() < expected_len {
                    out.push(val);
                }
            }
        }
    }

    out.resize(expected_len, 0);
    out
}

File: tools/willy_re/willy_re/director/bitmap.py
"""BITD (bitmap data) decoder for Director 5/6.

Supports 1-bit, 8-bit (paletted), 16-bit, and 32-bit bitmaps.
8-bit uses PackBits RLE compression; 1-bit is raw bit-packed.
"""

from __future__ import annotations

import logging
import struct
from typing import BinaryIO

from PIL import Image

log = logging.getLogger(__name__)


def decode_bitd(
    f: BinaryIO,
    offset: int,
    length: int,
    width: int,
    height: int,
    bit_depth: int,
) -> list[list]:
    """Decode a BITD chunk into a 2D pixel array.

    For 32-bit images, each pixel is [A, R, G, B].
    For 8-bit paletted, each pixel is a palette index (0-255).
    For 1-bit, each pixel is 0 or 1.

    Parameters
    ----------
    f : BinaryIO
        File object positioned at start of the BITD chunk data (after FourCC + length).
    offset : int
        Absolute file offset of the BITD entry (before FourCC).
    length : int
        Length of the BITD data.
    width, height : int
        Image dimensions from the CASt member.
    bit_depth : int
        Bit depth: 1, 8, 16, or 32.
    """
    if width <= 0 or height <= 0:
        return []

    f.seek(offset + 8)  # skip FourCC + length

    if bit_depth == 32:
        return _decode_32bit(f, offset, length, width, height)
    elif bit_depth >= 33:
        # 1-bit stored as bit_depth > 32 in Director quirk
        return _decode_1bit(f, offset, length, width, height)
    elif bit_depth == 16:
        return _decode_16bit(f, offset, length, width, height)
    elif bit_depth == 4:
        return _decode_4bit(f, offset, length, width, height)
    elif bit_depth == 2:
        return _decode_2bit(f, offset, length, width, height)
    else:
        return _decode_paletted(f, offset, length, width, height)


def _decode_2bit(f: BinaryIO, offset: int, length: int, width: int, height: int) -> list[list[int]]:
    """Decode 2-bit paletted image: 4 pixels per byte, MSB first."""
    pixels = [[0] * width for _ in range(height)]
    x, y = 0, 0

    end = offset + length + 8
    while f.tell() < end and y < height:
        byte = f.read(1)
        if not byte:
            break
        val = byte[0]
        for shift in (6, 4, 2, 0):
            if x < width:
                pixels[y][x] = 0x03 - ((val >> shift) & 0x03)
            x += 1
            if x >= width:
                x = 0
                y += 1
                if y >= height:
                    return pixels
    return pixels


def _decode_4bit(f: BinaryIO, offset: int, length: int, width: int, height: int) -> list[list[int]]:
    """Decode 4-bit paletted image: 2 pixels per byte, high nibble first."""
    pixels = [[0] * width for _ in range(height)]
    x, y = 0, 0

    end = offset + length + 8
    while f.tell() < end and y < height:
        byte = f.read(1)
        if not byte:
            break
        val = byte[0]
        for shift in (4, 0):
            if x < width:
                pixels[y][x] = 0x0F - ((val >> shift) & 0x0F)
            x += 1
            if x >= width:
                x = 0
                y += 1
                if y >= height:
                    return pixels
    return pixels


def _decode_1bit(f: BinaryIO, offset: int, length: int, width: int, height: int) -> list[list[int]]:
    """Decode 1-bit image: 8 pixels per byte, MSB first."""
    pixels = [[0] * width for _ in range(height)]
    x, y = 0, 0

    while f.tell() < offset + length + 8 and y < height:
        byte = f.read(1)
        if not byte:
            break
        val = byte[0]
        for bit in range(7, -1, -1):
            if x < width:
                pixels[y][x] = 1 - ((val >> bit) & 1)
            x += 1
            if x >= width:
                x = 0
                y += 1
                if y >= height:
                    return pixels
    return pixels


def _decode_16bit(
    f: BinaryIO, offset: int, length: int, width: int, height: int
) -> list[list[list[int]]]:
    """Decode 16-bit RGB555 image (1 or 0 padding bits, 5 bits per channel).

    Director 16-bit BMPs use RGB555 (or occasionally RGB565).
    Returns pixels as [R, G, B] lists.
    """
    row_bytes = width * 2
    # Rows are padded to even byte boundaries (already even for 16-bit)
    if row_bytes % 2:
        row_bytes += 1

    pixels = [[[0, 0, 0] for _ in range(width)] for _ in range(height)]

    # Check if data is RLE-compressed or raw
    raw_size = row_bytes * height
    if raw_size <= length:
        # Raw / uncompressed
        for y in range(height):
            for x in range(width):
                word_bytes = f.read(2)
                if len(word_bytes) < 2:
                    return pixels
                word = struct.unpack_from(">H", word_bytes)[0]
                r = ((word >> 10) & 0x1F) << 3
                g = ((word >> 5) & 0x1F) << 3
                b = (word & 0x1F) << 3
                pixels[y][x] = [r, g, b]
            # Skip row padding
            pad_bytes = row_bytes - width * 2
            if pad_bytes > 0:
                f.read(pad_bytes)
    else:
        # PackBits RLE — decode into raw 16-bit pixel buffer
        end = offset + length + 8
        x, y = 0, 0
        while f.tell() < end and y < height:
            raw = f.read(1)
            if not raw:
                break
            n = raw[0]

            if n == 0x80:
                continue  # PackBits no-op
            elif n > 0x80:
                # RLE: repeat next 2 bytes (0x101 - n) times
                run_len = 0x101 - n
                word_bytes = f.read(2)
                if len(word_bytes) < 2:
                    break
                word = struct.unpack_from(">H", word_bytes)[0]
                r = ((word >> 10) & 0x1F) << 3
                g = ((word >> 5) & 0x1F) << 3
                b = (word & 0x1F) << 3
                for _ in range(run_len):
                    if y >= height:
                        return pixels
                    pixels[y][x] = [r, g, b]
                    x += 1
                    if x >= width:
                        x = 0
                        y += 1
            else:
                # Literal: copy (n+1) 16-bit pixels
                copy_len = n + 1
                for _ in range(copy_len):
                    word_bytes = f.read(2)
                    if len(word_bytes) < 2:
                        break
                    if y >= height:
                        return pixels
                    word = struct.unpack_from(">H", word_bytes)[0]
                    r = ((word >> 10) & 0x1F) << 3
                    g = ((word >> 5) & 0x1F) << 3
                    b = (word & 0x1F) << 3
                    pixels[y][x] = [r, g, b]
                    x += 1
                    if x >= width:
                        x = 0
                        y += 1

    return pixels


def _decode_paletted(
    f: BinaryIO, offset: int, length: int, width: int, height: int
) -> list[list[int]]:
    """Decode 8-bit paletted image (PackBits RLE or raw)."""
    pixels = [[0] * width for _ in range(height)]
    x, y = 0, 0

    # Determine whether data is raw (uncompressed) or RLE
    pad = height if width % 2 else 0
    raw_size = width * height + pad

    if raw_size == length:
        # Direct / uncompressed mode
        return _decode_raw_paletted(f, offset, length, width, height, pad > 0)

    # RLE mode (PackBits)
    # Row stride includes a padding byte for odd widths
    row_stride = width + (width % 2)
    end = offset + length + 8
    while f.tell() < end and y < height:
        raw = f.read(1)
        if not raw:
            break
        n = raw[0]

        if n == 0x80:
            # PackBits no-op byte — skip
            continue
        elif n > 0x7F:
            # Run-length: repeat next byte (0x101 - n) times
            run_len = 0x101 - n
            val_byte = f.read(1)
            if not val_byte:
                break
            val = 0xFF - val_byte[0]
            for _ in range(run_len):
                if y >= height:
                    return pixels
                if x < width:
                    pixels[y][x] = val
                x += 1
                if x >= row_stride:
                    x = 0
                    y += 1
        else:
            # Literal: copy next (n + 1) bytes
            copy_len = n + 1
            for _ in range(copy_len):
                val_byte = f.read(1)
                if not val_byte:
                    break
                if y >= height:
                    return pixels
                if x < width:
                    pixels[y][x] = 0xFF - val_byte[0]
                x += 1
                if x >= row_stride:
                    x = 0
                    y += 1

    return pixels


def _decode_raw_paletted(
    f: BinaryIO, offset: int, length: int, width: int, height: int, has_pad: bool
) -> list[list[int]]:
    """Decode raw (uncompressed) 8-bit paletted image."""
    pixels = [[0] * width for _ in range(height)]
    for y in range(height):
        for x in range(width):
            b = f.read(1)
            if not b:
                return pixels
            pixels[y][x] = 0xFF - b[0]
        if has_pad:
            f.read(1)  # skip padding byte
    return pixels


def _decode_32bit(
    f: BinaryIO, offset: int, length: int, width: int, height: int
) -> list[list[list[int]]]:
    """Decode 32-bit ARGB image (PackBits RLE, channel-planar per row)."""
    pixels = [[[0, 0, 0, 255] for _ in range(width)] for _ in range(height)]
    x, y = 0, 0
    channel = 0  # 0=A, 1=R, 2=G, 3=B

    end = offset + length + 8
    while f.tell() < end and y < height:
        raw = f.read(1)
        if not raw:
            break
        n = raw[0]

        if n > 0x7F:
            run_len = 0x101 - n
            val_byte = f.read(1)
            if not val_byte:
                break
            val = val_byte[0]
            for _ in range(run_len):
                if y >= height:
                    return pixels
                pixels[y][x][channel] = val
                x += 1
                if x >= width:
                    channel = (channel + 1) % 4
                    x = 0
                    if channel == 0:
                        y += 1
        else:
            copy_len = n + 1
            for _ in range(copy_len):
                val_byte = f.read(1)
                if not val_byte:
                    break
                if y >= height:
                    return pixels
                pixels[y][x][channel] = val_byte[0]
                x += 1
                if x >= width:
                    channel = (channel + 1) % 4
                    x = 0
                    if channel == 0:
                        y += 1

    return pixels


# ---------------------------------------------------------------------------
# High-level: BITD → PIL Image
# ---------------------------------------------------------------------------


def bitd_to_image(
    f: BinaryIO,
    offset: int,
    length: int,
    width: int,
    height: int,
    bit_depth: int,
    palette: list[tuple[int, int, int]] | None = None,
    palette_id: int = 0,
    transparent_white: bool = True,
    is_windows: bool = False,
) -> Image.Image | None:
    """Decode BITD to a PIL Image.

    Parameters
    ----------
    palette : optional list of (R,G,B) tuples for 8-bit images.
    palette_id : built-in palette selector (negative = system palette).
    transparent_white : if True, white pixels become transparent (for sprite compositing).
    is_windows : if True, use Windows system palette as default instead of Mac.
    """
    if width <= 0 or height <= 0:
        return None

    pixels = decode_bitd(f, offset, length, width, height, bit_depth)
    if not pixels:
        return None

    if bit_depth == 32:
        img = Image.new("RGBA", (width, height))
        for y in range(height):
            for x in range(width):
                p = pixels[y][x]
                img.putpixel((x, y), (p[1], p[2], p[3], 255 - p[0]))
        return img

    elif bit_depth >= 33:
        # 1-bit
        img = Image.new("1", (width, height))
        for y in range(height):
            for x in range(width):
                img.putpixel((x, y), pixels[y][x])
        return img

    elif bit_depth == 16:
        # 16-bit RGB555 → RGB image
        img = Image.new("RGB", (width, height))
        for y in range(height):
            for x in range(width):
                p = pixels[y][x]
                img.putpixel((x, y), (p[0], p[1], p[2]))
        return img

    else:
        # 8-bit paletted
        img = Image.new("P", (width, height))
        flat_palette = _build_flat_palette(palette, palette_id, is_windows=is_windows)
        if flat_palette:
            img.putpalette(flat_palette)

        for y in range(height):
            for x in range(width):
                img.putpixel((x, y), pixels[y][x])

        if transparent_white:
            return img.convert("RGBA")
        return img


def _build_flat_palette(
    palette: list[tuple[int, int, int]] | None,
    palette_id: int,
    *,
    is_windows: bool = False,
) -> list[int] | None:
    """Build a flat [R,G,B,R,G,B,...] palette list for PIL."""
    if palette:
        flat: list[int] = []
        for r, g, b in palette:
            flat.extend([r, g, b])
        # Pad to 256 entries
        while len(flat) < 768:
            flat.extend([0, 0, 0])
        return flat

    # Use system palette — pick Windows default for XFIR files
    from .palette import get_system_palette

    effective_id = palette_id
    if effective_id == 0 and is_windows:
        effective_id = -100  # Windows system palette

    sys_pal = get_system_palette(effective_id)
    if sys_pal:
        flat = []
        for r, g, b in sys_pal:
            flat.extend([r, g, b])
        while len(flat) < 768:
            flat.extend([0, 0, 0])
        return flat

    return None
